_posts/2017-04-02-generating-sitemaps-with-apache-spark.markdown

```diff
@@ -0,0 +1,228 @@
+---
+layout: post
+title: "Generating Sitemaps with Apache Spark"
+date: 2017-04-02
+comments: true
+author: db
+categories: [Big Data, Apache Spark]
+---
+While Artsy is the largest database of Contemporary Art online, it's not exactly "big data". To date, we have published 604,897 artworks by 69,110 artists from 4,192 partners (galleries, museums, institutions) in 43,749 shows. Our team has written 8,778 articles, hosted 187 art fairs and a few dozen auctions. We have 1,015 genes from the Art Genome project.
+
+There're just over a million web pages generated from this data on artsy.net. Generating sitemaps to submit to Google and other search engines for a million pages never seemed like a big deal. In this post I'll describe 3 generations of code, including our most recent iteration that uses Apache Spark to generates static sitemap files in S3.
+
+<!-- more -->
+
+### What's in a Sitemap?
+
+If you're not familiar with sitemaps, they are an easy way for us to inform search engines about pages on artsy.net available for crawling. A Sitemap is an XML file that lists URLs along with some additional metadata. All Artsy sitemaps are listed in our [robots.txt](https://www.artsy.net/robots.txt).
+
+```
+User-agent: *
+Sitemap: https://www.artsy.net/sitemap-artists.xml
+Sitemap: https://www.artsy.net/sitemap-shows.xml
+...
+```
+
+These are actually sitemap indexes, such as [the shows sitemap index](https://www.artsy.net/sitemap-shows.xml). Each index contains links to the actual sitemaps.
+
+```xml
+<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+ <sitemap>
+  <loc>https://www.artsy.net/sitemap-shows-2016.xml</loc>
+ </sitemap>
+ ...
+</sitemapindex>
+```
+
+Each sitemap contains links to web pages on [www.artsy.net](https://www.artsy.net), such as the [sitemap containing links to shows in 2016](https://www.artsy.net/sitemap-shows-2016.xml).
+
+```xml
+<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
+ <url>
+  <loc>https://www.artsy.net/show/guy-hepner-kate-moss-exhibiton</loc>
+ </url>
+ ...
+</urlset>
+```
+
+### The Monolith's Rake Task
+
+When Artsy was a Ruby on Rails monolith we generated and wrote static sitemap files to S3 in a Rake task. We iterated over all our data nightly, created a set of URLs and wrote an XML file. We got fairly clever with a `Sitemappable` module that was included in anything that yielded a sitemap.
+
+```ruby
+module Sitemappable
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    def sitemappable_priority
+      0.5
+    end
+  end
+end
+```
+
+We also used HAML to generate sitemaps.
```

Comment:
This will send readers on an ill-advised research tangent. Maybe clarify that "We even used HAML, the same templating system we used to render web pages, to render sitemaps."